File: python/lib/woothee/crawler.py
# -*- coding: utf-8 -*-

import re
from . import dataset
from . import util

def challenge_google(ua, result):
  if 'Google' not in ua:
    return False
  
  if 'compatible; Googlebot' in ua:
    if 'compatible; Googlebot-Mobile' in ua:
      util.update_map(result, dataset.get('GoogleBotMobile'))
    else:
      util.update_map(result, dataset.get('GoogleBot'))
    return True
  
  if 'Googlebot-Image/' in ua:
    util.update_map(result, dataset.get('GoogleBot'))
    return True
  
  if 'Mediapartners-Google' in ua:
    if 'compatible; Mediapartners-Google' in ua or ua == 'Mediapartners-Google':
      util.update_map(result, dataset.get('GoogleMediaPartners'))
      return True
  
  if 'Feedfetcher-Google;' in ua:
    util.update_map(result, dataset.get('GoogleFeedFetcher'))
    return True
  
  if 'AppEngine-Google' in ua:
    util.update_map(result, dataset.get('GoogleAppEngine'))
    return True
  
  if 'Google Web Preview' in ua:
    util.update_map(result, dataset.get('GoogleWebPreview'))
    return True
  
  return False

def challenge_crawlers(ua, result):
  if 'Yahoo' in ua or 'help.yahoo.co.jp/help/jp/' in ua or 'listing.yahoo.co.jp/support/faq/' in ua:
    if 'compatible; Yahoo! Slurp' in ua:
      util.update_map(result, dataset.get('YahooSlurp'))
      return True
    
    if 'YahooFeedSeekerJp' in ua or 'YahooFeedSeekerBetaJp' in ua:
      util.update_map(result, dataset.get('YahooJP'))
      return True
    
    if 'crawler (http://listing.yahoo.co.jp/support/faq/' in ua or 'crawler (http://help.yahoo.co.jp/help/jp/' in ua:
      util.update_map(result, dataset.get('YahooJP'))
      return True
    
    if 'Yahoo Pipes' in ua:
      util.update_map(result, dataset.get('YahooPipes'))
      return True
  
  if 'msnbot' in ua:
    util.update_map(result, dataset.get('msnbot'))
    return True
  
  if 'bingbot' in ua:
    if 'compatible; bingbot' in ua:
      util.update_map(result, dataset.get('bingbot'))
      return True
  
  if 'Baidu' in ua:
    if 'compatible; Baiduspider' in ua or 'Baiduspider+' in ua or 'Baiduspider-image+' in ua:
      util.update_map(result, dataset.get('Baiduspider'))
      return True
    
  if 'Yeti' in ua:
    if 'http://help.naver.com/robots' in ua:
      util.update_map(result, dataset.get('Yeti'))
      return True
  
  if 'FeedBurner/' in ua:
    util.update_map(result, dataset.get('FeedBurner'))
    return True
  
  if 'facebookexternalhit' in ua:
    util.update_map(result, dataset.get('facebook'))
    return True
  
  if 'ichiro' in ua:
    if 'http://help.goo.ne.jp/door/crawler.html' in ua or 'compatible; ichiro/mobile goo;' in ua:
      util.update_map(result, dataset.get('goo'))
      return True
  if 'gooblogsearch/' in ua:
    util.update_map(result, dataset.get('goo'))
    return True
  
  if 'Apple-PubSub' in ua:
    util.update_map(result, dataset.get('ApplePubSub'))
    return True

  if '(www.radian6.com/crawler)' in ua:
    util.update_map(result, dataset.get('radian6'))
    return True

  if 'Genieo/' in ua:
    util.update_map(result, dataset.get('Genieo'))
    return True

  if 'labs.topsy.com/butterfly/' in ua:
    util.update_map(result, dataset.get('topsyButterfly'))
    return True

  if 'rogerbot/1.0 (http://www.seomoz.org/dp/rogerbot' in ua:
    util.update_map(result, dataset.get('rogerbot'))
    return True

  if 'compatible; AhrefsBot/' in ua:
    util.update_map(result, dataset.get('AhrefsBot'))
    return True
  
  if 'livedoor FeedFetcher' in ua or 'Fastladder FeedFetcher' in ua:
    util.update_map(result, dataset.get('livedoorFeedFetcher'))
    return True
  
  if 'Hatena ' in ua:
    if 'Hatena Antenna' in ua or 'Hatena Pagetitle Agent' in ua or 'Hatena Diary RSS' in ua:
      util.update_map(result, dataset.get('Hatena'))
      return True
  
  if 'mixi-check' in ua or 'mixi-crawler' in ua or 'mixi-news-crawler' in ua:
    util.update_map(result, dataset.get('mixi'))
    return True
  
  if 'Indy Library' in ua:
    if 'compatible; Indy Library' in ua:
      util.update_map(result, dataset.get('IndyLibrary'))
      return True
  
  return False

def challenge_maybe_crawler(ua, result):
  if re.search('(bot|crawler|spider)(?:[-_ ./;@()]|$)', ua, re.I):
    util.update_map(result, dataset.get('VariousCrawler'))
    return True
  
  if re.search('(?:Rome Client |UnwindFetchor/|ia_archiver |Summify |PostRank/)', ua) or 'ASP-Ranker Feed Crawler' in ua:
    util.update_map(result, dataset.get('VariousCrawler'))
    return True
  
  if re.search('(feed|web) ?parser', ua, re.I):
    util.update_map(result, dataset.get('VariousCrawler'))
    return True
  
  if re.search('watch ?dog', ua, re.I):
    util.update_map(result, dataset.get('VariousCrawler'))
    return True
  
  return False

File: ruby/lib/woothee/crawler.rb
# -*- coding: utf-8 -*-

require 'woothee/dataset'
require 'woothee/util'

module Woothee::Crawler
  extend Woothee::Util

  def self.challenge_google(ua, result)
    return false if ua.index('Google').nil?

    if ua.index('compatible; Googlebot')
      if ua.index('compatible; Googlebot-Mobile')
        update_map(result, Woothee::DataSet.get('GoogleBotMobile'))
        return true
      else
        update_map(result, Woothee::DataSet.get('GoogleBot'))
        return true
      end
    end
    if ua.index('Googlebot-Image/')
      update_map(result, Woothee::DataSet.get('GoogleBot'))
      return true
    end
    if ua.index('Mediapartners-Google')
      if ua.index('compatible; Mediapartners-Google') or ua == 'Mediapartners-Google'
        update_map(result, Woothee::DataSet.get('GoogleMediaPartners'))
        return true
      end
    end
    if ua.index('Feedfetcher-Google;')
      update_map(result, Woothee::DataSet.get('GoogleFeedFetcher'))
      return true
    end
    if ua.index('AppEngine-Google')
      update_map(result, Woothee::DataSet.get('GoogleAppEngine'))
      return true
    end
    if ua.index('Google Web Preview')
      update_map(result, Woothee::DataSet.get('GoogleWebPreview'))
      return true
    end

    false
  end

  def self.challenge_crawlers(ua, result)
    if ua.index('Yahoo') or ua.index('help.yahoo.co.jp/help/jp/') or ua.index('listing.yahoo.co.jp/support/faq/')
      if ua.index('compatible; Yahoo! Slurp')
        update_map(result, Woothee::DataSet.get('YahooSlurp'))
        return true
      end
      if ua.index('YahooFeedSeekerJp') or ua.index('YahooFeedSeekerBetaJp')
        update_map(result, Woothee::DataSet.get('YahooJP'))
        return true
      end
      if ua.index('crawler (http://listing.yahoo.co.jp/support/faq/') || ua.index('crawler (http://help.yahoo.co.jp/help/jp/')
        update_map(result, Woothee::DataSet.get('YahooJP'))
        return true
      end
      if ua.index('Yahoo Pipes')
        update_map(result, Woothee::DataSet.get('YahooPipes'))
        return true
      end
    end
    if ua.index('msnbot')
      update_map(result, Woothee::DataSet.get('msnbot'))
      return true
    end
    if ua.index('bingbot')
      if ua.index('compatible; bingbot')
        update_map(result, Woothee::DataSet.get('bingbot'))
        return true
      end
    end
    if ua.index('Baidu')
      if ua.index('compatible; Baiduspider') or ua.index('Baiduspider+') or ua.index('Baiduspider-image+')
        update_map(result, Woothee::DataSet.get('Baiduspider'))
        return true
      end
    end
    if ua.index('Yeti')
      if ua.index('http://help.naver.com/robots')
        update_map(result, Woothee::DataSet.get('Yeti'))
        return true
      end
    end
    if ua.index('FeedBurner/')
      update_map(result, Woothee::DataSet.get('FeedBurner'))
      return true
    end
    if ua.index('facebookexternalhit')
      update_map(result, Woothee::DataSet.get('facebook'))
      return true
    end
    if ua.index('ichiro')
      if ua.index('http://help.goo.ne.jp/door/crawler.html') or ua.index('compatible; ichiro/mobile goo;')
        update_map(result, Woothee::DataSet.get('goo'))
        return true
      end
    end
    if ua.index('gooblogsearch/')
      update_map(result, Woothee::DataSet.get('goo'))
      return true
    end
    if ua.index('Apple-PubSub')
      update_map(result, Woothee::DataSet.get('ApplePubSub'))
      return true
    end
    if ua.index("(www.radian6.com/crawler)")
      update_map(result, Woothee::DataSet.get("radian6"))
      return true
    end
    if ua.index('Genieo/')
      update_map(result, Woothee::DataSet.get("Genieo"))
      return true
    end
    if ua.index("labs.topsy.com/butterfly/")
      update_map(result, Woothee::DataSet.get("topsyButterfly"))
      return true
    end
    if ua.index("rogerbot/1.0 (http://www.seomoz.org/dp/rogerbot")
      update_map(result, Woothee::DataSet.get("rogerbot"))
      return true
    end
    if ua.index("compatible; AhrefsBot/")
      update_map(result, Woothee::DataSet.get("AhrefsBot"))
      return true
    end
    if ua.index('livedoor FeedFetcher') or ua.index('Fastladder FeedFetcher')
      update_map(result, Woothee::DataSet.get('livedoorFeedFetcher'))
      return true
    end
    if ua.index('Hatena ')
      if ua.index('Hatena Antenna') or ua.index('Hatena Pagetitle Agent') or ua.index('Hatena Diary RSS')
        update_map(result, Woothee::DataSet.get('Hatena'))
        return true
      end
    end
    if ua.index('mixi-check') or ua.index('mixi-crawler') or ua.index('mixi-news-crawler')
      update_map(result, Woothee::DataSet.get('mixi'))
      return true
    end
    if ua.index('Indy Library')
      if ua.index('compatible; Indy Library')
        update_map(result, Woothee::DataSet.get('IndyLibrary'))
        return true
      end
    end

    false
  end

  def self.challenge_maybe_crawler(ua, result)
    if ua =~ /(bot|crawler|spider)(?:[-_ .\/;@()]|$)/oi
      update_map(result, Woothee::DataSet.get('VariousCrawler'))
      return true
    end

    if ua =~ /(?:Rome Client |UnwindFetchor\/|ia_archiver |Summify |PostRank\/)/o or ua.index('ASP-Ranker Feed Crawler')
      update_map(result, Woothee::DataSet.get('VariousCrawler'))
      return true
    end

    if ua =~ /(feed|web) ?parser/oi
      update_map(result, Woothee::DataSet.get('VariousCrawler'))
      return true
    end

    if ua =~ /watch ?dog/oi
      update_map(result, Woothee::DataSet.get('VariousCrawler'))
      return true
    end

    false
  end
end
